src/bin/pg_autoctl/supervisor.c

```diff
@@ -0,0 +1,827 @@
+				{
+					++subprocessCount;
+				}
+
+				break;
+			}
+		}
+
+		/* avoid buzy looping on waitpid(WNOHANG) */
+		pg_usleep(100 * 1000); /* 100 ms */
+	}
+
+	/* we track in the main loop if it's a cleanExit or not */
+	return supervisor->cleanExit;
+}
+
+
+/*
+ * supervisor_find_service loops over the SubProcess array to find given pid and
+ * return its entry in the array.
+ */
+static bool
+supervisor_find_service(Supervisor *supervisor, pid_t pid, Service **result)
+{
+	int serviceCount = supervisor->serviceCount;
+	int serviceIndex = 0;
+
+	for (serviceIndex = 0; serviceIndex < serviceCount; serviceIndex++)
+	{
+		if (pid == supervisor->services[serviceIndex].pid)
+		{
+			*result = &(supervisor->services[serviceIndex]);
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * supervisor_reload_services sends SIGHUP to all our services.
+ */
+static void
+supervisor_reload_services(Supervisor *supervisor)
+{
+	int serviceCount = supervisor->serviceCount;
+	int serviceIndex = 0;
+
+	for (serviceIndex = 0; serviceIndex < serviceCount; serviceIndex++)
+	{
+		Service *service = &(supervisor->services[serviceIndex]);
+
+		log_info("Reloading service \"%s\" by signaling pid %d with SIGHUP",
+				 service->name, service->pid);
+
+		if (kill(service->pid, SIGHUP) != 0)
+		{
+			log_error("Failed to send SIGQUIT to service %s with pid %d",
+					  service->name, service->pid);
+		}
+	}
+
+	/* reset our signal handling facility */
+	asked_to_reload = 0;
+}
+
+
+/*
+ * supervisor_stop_subprocesses calls the stopFunction for all the registered
+ * services to initiate the shutdown sequence.
+ */
+static void
+supervisor_stop_subprocesses(Supervisor *supervisor)
+{
+	int signal = asked_to_stop_fast ? SIGINT : SIGTERM;
+	int serviceCount = supervisor->serviceCount;
+	int serviceIndex = 0;
+
+	for (serviceIndex = 0; serviceIndex < serviceCount; serviceIndex++)
+	{
+		Service *service = &(supervisor->services[serviceIndex]);
+
+		if (kill(service->pid, signal) != 0)
+		{
+			log_error("Failed to send signal %s to service %s with pid %d",
+					  strsignal(signal), service->name, service->pid);
+		}
+	}
+}
+
+
+/*
+ * supervisor_stop_other_subprocesses sends the QUIT signal to other known
+ * sub-processes when on of does is reported dead.
+ */
+static void
+supervisor_stop_other_services(Supervisor *supervisor, pid_t pid)
+{
+	int signal = asked_to_stop_fast ? SIGINT : SIGTERM;
+	int serviceCount = supervisor->serviceCount;
+	int serviceIndex = 0;
+
+	/*
+	 * In case of unexpected stop (bug), we stop the other processes too.
+	 * Someone might then notice (such as systemd) and restart the whole
+	 * thing again.
+	 */
+	if (!(asked_to_stop || asked_to_stop_fast))
+	{
+		for (serviceIndex = 0; serviceIndex < serviceCount; serviceIndex++)
+		{
+			Service *service = &(supervisor->services[serviceIndex]);
+
+			if (service->pid != pid)
+			{
+				if (kill(service->pid, signal) != 0)
+				{
+					log_error("Failed to send SIGQUIT to service %s with pid %d",
```

Comment:
Should use `strsignal` not to replace `SIGQUIT` here.

Reply:
Done.